File: sdks/python/apex_sdk/pagination.py
"""Auto-iterates a cursor-paginated list endpoint, yielding one item at a time
and fetching the next page on demand."""

from __future__ import annotations

from typing import Any, Callable, Iterator

from .types import Page


def paginate_all(fetch_page: Callable[..., Page], **params: Any) -> Iterator[Any]:
    """Works with any resource method shaped like `(*, cursor=None, **params)
    -> Page`, e.g. `paginate_all(client.agents.list, limit=25)`. Stops once
    `has_more` is false or `next_cursor` is `None`, whichever comes first."""
    cursor = None
    while True:
        page = fetch_page(cursor=cursor, **params)
        for item in page["data"]:
            yield item
        if not page["has_more"] or not page["next_cursor"]:
            return
        cursor = page["next_cursor"]

File: sdks/typescript/src/pagination.ts
import type { Page, PageParams } from "./types.js";

/** Auto-iterates a cursor-paginated list endpoint, yielding one item at a
 * time and fetching the next page on demand. Works with any resource method
 * shaped like `(params?: PageParams) => Promise<Page<T>>` — e.g.
 * `paginateAll(params => client.agents.list(params))`. Stops once
 * `has_more` is `false` or `next_cursor` is `null`, whichever comes first. */
export async function* paginateAll<T, P extends PageParams = PageParams>(
  fetchPage: (params: P) => Promise<Page<T>>,
  params?: Omit<P, "cursor">,
): AsyncGenerator<T> {
  let cursor: string | undefined;
  for (;;) {
    const page = await fetchPage({ ...(params as P), cursor });
    for (const item of page.data) {
      yield item;
    }
    if (!page.has_more || !page.next_cursor) return;
    cursor = page.next_cursor;
  }
}
